File: sdk/typescript/src/namespaces/belief.ts
/**
 * Belief Network Namespace API
 *
 * Provides access to belief network analysis, cruxes, and provenance tracking.
 * The belief network represents claims and their supporting/opposing relationships.
 */

import type { AragoraClient } from '../client';

/**
 * Crux analysis result from CruxDetector.
 */
export interface CruxAnalysis {
  debate_id: string;
  cruxes: Array<{
    claim_id: string;
    statement: string;
    influence_score: number;
    disagreement_score: number;
    uncertainty_score: number;
    centrality_score: number;
    composite_score: number;
  }>;
  count: number;
}

/**
 * Belief crux representing a key claim that would impact debate outcome.
 */
export interface BeliefCrux {
  claim_id: string;
  claim_text: string;
  impact_score: number;
  current_confidence: number;
  supporting_agents: string[];
  opposing_agents: string[];
}

/**
 * Response for cruxes endpoint.
 */
export interface BeliefCruxes {
  debate_id: string;
  cruxes: BeliefCrux[];
  analysis_timestamp: string;
}

/**
 * Load-bearing claim with centrality score.
 */
export interface LoadBearingClaim {
  claim_id: string;
  claim_text: string;
  centrality_score: number;
  dependents_count: number;
}

/**
 * Response for load-bearing claims.
 */
export interface LoadBearingClaims {
  debate_id: string;
  claims: LoadBearingClaim[];
}

/**
 * Node in the belief graph.
 */
export interface BeliefGraphNode {
  id: string;
  text: string;
  type: 'claim' | 'evidence' | 'inference';
  confidence: number;
  is_crux?: boolean;
}

/**
 * Link between nodes in the belief graph.
 */
export interface BeliefGraphLink {
  source: string;
  target: string;
  type: 'supports' | 'opposes' | 'depends_on';
  strength: number;
}

/**
 * Complete belief graph structure.
 */
export interface BeliefGraph {
  debate_id: string;
  nodes: BeliefGraphNode[];
  links: BeliefGraphLink[];
}

/**
 * Export format options.
 */
export type ExportFormat = 'json' | 'graphml' | 'csv';

/**
 * Exported belief network.
 */
export interface BeliefExport {
  format: ExportFormat;
  content: string;
  metadata: {
    debate_id: string;
    exported_at: string;
    node_count: number;
    link_count: number;
  };
}

/**
 * Belief Network API.
 *
 * Provides methods for analyzing belief networks in debates:
 * - Finding cruxes (key claims that would change outcomes)
 * - Identifying load-bearing claims (high centrality)
 * - Exporting graph structures
 */
export class BeliefAPI {
  constructor(private client: AragoraClient) {}

  /**
   * Get key claims (cruxes) that would most impact the debate outcome.
   * @route GET /api/v1/belief-network/{debate_id}/cruxes
   */
  async getCruxes(
    debateId: string,
    options?: { top_k?: number }
  ): Promise<BeliefCruxes> {
    return this.client.request(
      'GET',
      `/api/v1/belief-network/${encodeURIComponent(debateId)}/cruxes`,
      { params: options }
    ) as Promise<BeliefCruxes>;
  }

  /**
   * Get claims with highest centrality (most load-bearing).
   * @route GET /api/v1/belief-network/{debate_id}/load-bearing-claims
   */
  async getLoadBearingClaims(
    debateId: string,
    options?: { limit?: number }
  ): Promise<LoadBearingClaims> {
    return this.client.request(
      'GET',
      `/api/v1/belief-network/${encodeURIComponent(debateId)}/load-bearing-claims`,
      { params: options }
    ) as Promise<LoadBearingClaims>;
  }

  /**
   * Get belief network as a graph structure for visualization.
   * @route GET /api/v1/belief-network/{debate_id}/graph
   */
  async getGraph(
    debateId: string,
    options?: { include_cruxes?: boolean }
  ): Promise<BeliefGraph> {
    return this.client.request(
      'GET',
      `/api/v1/belief-network/${encodeURIComponent(debateId)}/graph`,
      { params: options }
    ) as Promise<BeliefGraph>;
  }

  /**
   * Export belief network in various formats.
   * @route GET /api/v1/belief-network/{debate_id}/export
   */
  async export(
    debateId: string,
    options?: { format?: ExportFormat }
  ): Promise<BeliefExport> {
    return this.client.request(
      'GET',
      `/api/v1/belief-network/${encodeURIComponent(debateId)}/export`,
      { params: options }
    ) as Promise<BeliefExport>;
  }
}

File: sdk/python/aragora_sdk/namespaces/belief.py
"""
Belief Network Namespace API

Provides access to belief network analysis, cruxes, and provenance tracking.
"""

from __future__ import annotations

from typing import TYPE_CHECKING, Any, Literal

if TYPE_CHECKING:
    from ..client import AragoraAsyncClient, AragoraClient


class BeliefAPI:
    """Synchronous Belief Network API for claim analysis and provenance."""

    def __init__(self, client: AragoraClient):
        self._client = client

    def get_cruxes(
        self,
        debate_id: str,
        top_k: int = 3,
    ) -> dict[str, Any]:
        """Get key claims that would most impact the debate outcome.

        Args:
            debate_id: The debate ID.
            top_k: Number of top cruxes to return (1-10, default: 3).

        Returns:
            Cruxes with their impact scores and claim details.
        """
        params = {"top_k": top_k}
        return self._client.request(
            "GET", f"/api/v1/belief-network/{debate_id}/cruxes", params=params
        )

    def get_load_bearing_claims(
        self,
        debate_id: str,
        limit: int = 5,
    ) -> dict[str, Any]:
        """Get claims with highest centrality (most load-bearing).

        Args:
            debate_id: The debate ID.
            limit: Maximum claims to return (1-20, default: 5).

        Returns:
            Load-bearing claims with centrality scores.
        """
        params = {"limit": limit}
        return self._client.request(
            "GET", f"/api/v1/belief-network/{debate_id}/load-bearing-claims", params=params
        )

    def get_graph(
        self,
        debate_id: str,
        include_cruxes: bool = True,
    ) -> dict[str, Any]:
        """Get belief network as a graph structure for visualization.

        Args:
            debate_id: The debate ID.
            include_cruxes: Whether to include crux detection (default: True).

        Returns:
            Graph structure with nodes (claims) and links (influences).
        """
        params = {"include_cruxes": str(include_cruxes).lower()}
        return self._client.request(
            "GET", f"/api/v1/belief-network/{debate_id}/graph", params=params
        )

    def export(
        self,
        debate_id: str,
        format: Literal["json", "graphml", "csv"] = "json",
    ) -> dict[str, Any]:
        """Export belief network in various formats.

        Args:
            debate_id: The debate ID.
            format: Export format (json, graphml, csv).

        Returns:
            Exported belief network in requested format.
        """
        params = {"format": format}
        return self._client.request(
            "GET", f"/api/v1/belief-network/{debate_id}/export", params=params
        )

    def get_claim_support(
        self,
        debate_id: str,
        claim_id: str,
    ) -> dict[str, Any]:
        """Get verification status of all evidence supporting a claim.

        Args:
            debate_id: The debate ID.
            claim_id: The claim ID.

        Returns:
            Claim support status and evidence details.
        """
        return self._client.request(
            "GET", f"/api/v1/provenance/{debate_id}/claims/{claim_id}/support"
        )

    def get_graph_stats(
        self,
        debate_id: str,
    ) -> dict[str, Any]:
        """Get argument graph statistics for a debate.

        Args:
            debate_id: The debate ID.

        Returns:
            Graph statistics including node counts, edge counts, and density.
        """
        return self._client.request("GET", f"/api/v1/debate/{debate_id}/graph-stats")


class AsyncBeliefAPI:
    """Asynchronous Belief Network API for claim analysis and provenance."""

    def __init__(self, client: AragoraAsyncClient):
        self._client = client

    async def get_cruxes(
        self,
        debate_id: str,
        top_k: int = 3,
    ) -> dict[str, Any]:
        """Get key claims that would most impact the debate outcome."""
        params = {"top_k": top_k}
        return await self._client.request(
            "GET", f"/api/v1/belief-network/{debate_id}/cruxes", params=params
        )

    async def get_load_bearing_claims(
        self,
        debate_id: str,
        limit: int = 5,
    ) -> dict[str, Any]:
        """Get claims with highest centrality (most load-bearing)."""
        params = {"limit": limit}
        return await self._client.request(
            "GET", f"/api/v1/belief-network/{debate_id}/load-bearing-claims", params=params
        )

    async def get_graph(
        self,
        debate_id: str,
        include_cruxes: bool = True,
    ) -> dict[str, Any]:
        """Get belief network as a graph structure for visualization."""
        params = {"include_cruxes": str(include_cruxes).lower()}
        return await self._client.request(
            "GET", f"/api/v1/belief-network/{debate_id}/graph", params=params
        )

    async def export(
        self,
        debate_id: str,
        format: Literal["json", "graphml", "csv"] = "json",
    ) -> dict[str, Any]:
        """Export belief network in various formats."""
        params = {"format": format}
        return await self._client.request(
            "GET", f"/api/v1/belief-network/{debate_id}/export", params=params
        )

    async def get_claim_support(
        self,
        debate_id: str,
        claim_id: str,
    ) -> dict[str, Any]:
        """Get verification status of all evidence supporting a claim."""
        return await self._client.request(
            "GET", f"/api/v1/provenance/{debate_id}/claims/{claim_id}/support"
        )

    async def get_graph_stats(
        self,
        debate_id: str,
    ) -> dict[str, Any]:
        """Get argument graph statistics for a debate."""
        return await self._client.request("GET", f"/api/v1/debate/{debate_id}/graph-stats")
